published/external/xdpapi_helper.h

```diff
@@ -0,0 +1,229 @@
+
+#pragma once
+
+#include "xdpapi.h"
+
+#if defined(_KERNEL_MODE)
+
+extern XDP_API_PROVIDER_DISPATCH *XdpApiTest;
+extern XDP_API_PROVIDER_BINDING_CONTEXT *ProviderBindingContext;
+
+#define XDP_CALL_WITH_RETURN(ReturnType, FunctionCall, ...) \
+    do { \
+        ((XDP_API_CLIENT *)(ProviderBindingContext))->InProgressCallCount++; \
+        ReturnType result = XdpApiTest->FunctionCall(__VA_ARGS__); \
+        ((XDP_API_CLIENT *)(ProviderBindingContext))->InProgressCallCount--; \
+        return result; \
+    } while (0)
+
+#define XDP_CALL_VOID(FunctionCall, ...) \
+    do { \
+        ((XDP_API_CLIENT *)(ProviderBindingContext))->InProgressCallCount++; \
+        XdpApiTest->FunctionCall(__VA_ARGS__); \
+        ((XDP_API_CLIENT *)(ProviderBindingContext))->InProgressCallCount--; \
+    } while (0)
+
+// _IRQL_requires_(PASSIVE_LEVEL)
+XDP_STATUS
+CxPlatCreateProgram(
+    _In_ UINT32 InterfaceIndex,
+    _In_ const XDP_HOOK_ID *HookId,
+    _In_ UINT32 QueueId,
+    _In_ XDP_CREATE_PROGRAM_FLAGS Flags,
+    _In_reads_(RuleCount) const XDP_RULE *Rules,
+    _In_ UINT32 RuleCount,
+    _Out_ HANDLE *Program
+    )
+{
+    XDP_CALL_WITH_RETURN(XDP_STATUS, XdpCreateProgram, ProviderBindingContext, InterfaceIndex, HookId, QueueId, Flags, Rules, RuleCount, Program);
+}
+
+_IRQL_requires_(PASSIVE_LEVEL)
+VOID
+CxPlatXdpDeleteProgram(
+    _In_ HANDLE Program
+    )
+{
+    XDP_CALL_VOID(XdpDeleteProgram, Program);
+}
+
+_IRQL_requires_(PASSIVE_LEVEL)
+XDP_STATUS
+CxPlatXdpInterfaceOpen(
+    _In_ UINT32 InterfaceIndex,
+    _Out_ HANDLE *InterfaceHandle
+    )
+{
+    XDP_CALL_WITH_RETURN(XDP_STATUS, XdpInterfaceOpen, ProviderBindingContext, InterfaceIndex, InterfaceHandle);
+}
+
+_IRQL_requires_(PASSIVE_LEVEL)
+VOID
+CxPlatInterfaceClose(
+    _In_ HANDLE InterfaceHandle
+    )
+{
+    XDP_CALL_VOID(XdpInterfaceClose, InterfaceHandle);
+}
+
+_IRQL_requires_(PASSIVE_LEVEL)
+XDP_STATUS
+CxPlatXskCreate(
+    _In_opt_ PEPROCESS OwningProcess,
+    _In_opt_ PETHREAD OwningThread,
+    _In_opt_ PSECURITY_DESCRIPTOR SecurityDescriptor,
+    _Out_ HANDLE *Socket
+    )
+{
+    XDP_CALL_WITH_RETURN(XDP_STATUS, XskCreate, ProviderBindingContext, OwningProcess, OwningThread, SecurityDescriptor, Socket);
+}
+
+_IRQL_requires_(PASSIVE_LEVEL)
+VOID
+CxPlatXskDelete(
+    _In_ HANDLE Socket
+    )
+{
+    XDP_CALL_VOID(XskDelete, Socket);
+}
+
+XDP_STATUS
+CxPlatXskNotifyAsync2(
+    _In_ HANDLE Socket,
+    _In_ XSK_NOTIFY_FLAGS Flags,
+    _In_opt_ XSK_COMPLETION_CONTEXT CompletionContext,
+    _Out_ XSK_NOTIFY_RESULT_FLAGS *Result
+    )
+{
+    XDP_CALL_WITH_RETURN(XDP_STATUS, XskNotifyAsync2, Socket, Flags, CompletionContext, Result);
```

Comment:
One thing we should consider: it isn't necessary to keep track of every single downcall into the NMR provider if we can implicitly hold a reference via, e.g., handle creations. So I would implement this as follows:

1. ` XdpCreateProgram` and similar handle creation functions acquire a rundown reference. The functions do _not_ release the ref if handle creation succeeds.
2. All sockopt/notify etc. functions do not need to acquire/release refs because the handle param holds a ref.
3. When `XdpCloseHandle` wrapper function returns, it releases the original rundown ref.